File: packages/core/src/standing.ts
/**
 * Standing math (docs/trust-and-reach.md §4): the one server-computed,
 * server-authoritative enforcement factor. Pure functions only — no record
 * shapes, no signing, no floats-in-signed-records concern (this is math, not
 * a wire record). Clients never recompute standing itself (its inputs,
 * reports, are private by design); this module exists so the Go server and
 * the simlab share exactly the reference implementation the vectors pin down.
 */
import { CONSTANTS } from "./constants.js";

function clamp01(x: number): number {
  return Math.min(1, Math.max(0, x));
}

/**
 * p_adj(t) = p0 · 2^(-Δt / half-life) (trust-and-reach §4): the human rung
 * decays toward 0 with the standing half-life. p0 is clamped into [0,1]
 * first; a negative elapsedDays (clock skew, same-instant adjudication) is
 * treated as 0 elapsed.
 */
export function decayPenalty(
  p0: number,
  elapsedDays: number,
  halfLifeDays: number = CONSTANTS.standing_half_life_days,
): number {
  const p = clamp01(p0);
  const dt = Math.max(0, elapsedDays);
  return p * 2 ** (-dt / halfLifeDays);
}

/**
 * Reporter weight w(R) = (1 - p_adj(R)) × ln(1 + inbound_trust(R))
 * (trust-and-reach §4) — natural log, same scale as the budget formula
 * (budgets.ts). Deliberately the reporter's *adjudicated* component only
 * (not full standing including p_auto): discounting by p_auto would let
 * mass-reporting a target's likely defenders silence their reports
 * (report-the-reporters), and the definition would be recursive. Only a
 * human-confirmed false-report burn or uphold ever reduces a reporter's
 * future report weight.
 */
export function reporterWeight(adjPenalty: number, inboundTrust: number): number {
  return (1 - clamp01(adjPenalty)) * Math.log(1 + Math.max(0, inboundTrust));
}

function jaccard(a: ReadonlySet<string>, b: ReadonlySet<string>): number {
  let intersection = 0;
  for (const x of a) if (b.has(x)) intersection++;
  const union = a.size + b.size - intersection;
  // Jaccard(∅,∅) = 0 by convention here, never 1: two accounts with empty
  // outbound follow sets share no evidence of coordination and must NOT be
  // linked just because both sets are empty.
  if (union === 0) return 0;
  return intersection / union;
}

/**
 * Connected components of the reporter link graph (trust-and-reach §4):
 * reporters R1, R2 are linked iff either follows the other, or the Jaccard
 * overlap of their outbound follow sets >= jaccardThreshold. Deterministic
 * output: components sorted by their lexicographically-smallest member,
 * members sorted within each component. `reporters` is deduped defensively.
 */
export function clusterReporters(
  reporters: readonly string[],
  follows: Record<string, readonly string[]>,
  jaccardThreshold: number = CONSTANTS.report_cluster_jaccard,
): string[][] {
  const nodes = Array.from(new Set(reporters));
  const followSets = new Map<string, Set<string>>();
  for (const r of nodes) followSets.set(r, new Set(follows[r] ?? []));

  // Union-find over the deduped reporter list.
  const parent = new Map<string, string>();
  for (const r of nodes) parent.set(r, r);
  function find(x: string): string {
    let root = x;
    while (parent.get(root) !== root) root = parent.get(root)!;
    let cur = x;
    while (parent.get(cur) !== root) {
      const next = parent.get(cur)!;
      parent.set(cur, root);
      cur = next;
    }
    return root;
  }
  function union(a: string, b: string): void {
    const ra = find(a);
    const rb = find(b);
    if (ra !== rb) parent.set(ra, rb);
  }

  for (let i = 0; i < nodes.length; i++) {
    for (let j = i + 1; j < nodes.length; j++) {
      const a = nodes[i]!;
      const b = nodes[j]!;
      const aFollowsB = followSets.get(a)!.has(b);
      const bFollowsA = followSets.get(b)!.has(a);
      const linked =
        aFollowsB || bFollowsA || jaccard(followSets.get(a)!, followSets.get(b)!) >= jaccardThreshold;
      if (linked) union(a, b);
    }
  }

  const groups = new Map<string, string[]>();
  for (const r of nodes) {
    const root = find(r);
    const group = groups.get(root);
    if (group) group.push(r);
    else groups.set(root, [r]);
  }

  const components = Array.from(groups.values()).map((members) => [...members].sort());
  components.sort((a, b) => (a[0]! < b[0]! ? -1 : a[0]! > b[0]! ? 1 : 0));
  return components;
}

/**
 * Diversity-weighted report mass (trust-and-reach §4): each cluster
 * contributes only its maximum member weight — volume inside a cluster adds
 * nothing. Missing weight (a reporter absent from `weights`) counts as 0.
 */
export function reportMass(
  clusters: readonly (readonly string[])[],
  weights: Record<string, number>,
): number {
  let mass = 0;
  for (const cluster of clusters) {
    let max = 0;
    for (const member of cluster) {
      const w = weights[member] ?? 0;
      if (w > max) max = w;
    }
    mass += max;
  }
  return mass;
}

/** p_auto = min(cap, impact × max(0, mass)) (trust-and-reach §4). */
export function autoPenalty(
  mass: number,
  impact: number = CONSTANTS.report_impact,
  cap: number = CONSTANTS.report_auto_cap,
): number {
  return Math.min(cap, impact * Math.max(0, mass));
}

/** standing = (1 - p_auto) × (1 - p_adj) (protocol §9.3), each clamped to [0,1]. */
export function standingFrom(pAuto: number, pAdj: number): number {
  return (1 - clamp01(pAuto)) * (1 - clamp01(pAdj));
}

File: server/internal/trust/standing.go
package trust

import (
	"math"
	"sort"
)

// Standing math (docs/trust-and-reach.md §4): the one server-computed,
// server-authoritative enforcement factor. Pure functions only — this file
// mirrors packages/core/src/standing.ts function-for-function, and the
// shared standing-01 vector asserts the two agree to 1e-9. Clients never
// recompute standing itself (its inputs, reports, are private by design);
// the duplication exists so the server's enforcement and the simlab's
// simulation run the same published formulas.

func clamp01(x float64) float64 {
	return math.Min(1, math.Max(0, x))
}

// DecayPenalty is p_adj(t) = p0 · 2^(-Δt / half-life) (trust-and-reach §4):
// the human rung decays toward 0 with the standing half-life. p0 is clamped
// into [0,1] first; a negative elapsedDays (clock skew, same-instant
// adjudication) is treated as 0 elapsed.
func DecayPenalty(p0, elapsedDays, halfLifeDays float64) float64 {
	p := clamp01(p0)
	dt := math.Max(0, elapsedDays)
	return p * math.Pow(2, -dt/halfLifeDays)
}

// ReporterWeight is w(R) = (1 - p_adj(R)) × ln(1 + inbound_trust(R))
// (trust-and-reach §4) — natural log, the same scale as the budget formula
// (DailyBudget). Deliberately the reporter's *adjudicated* component only
// (not full standing including p_auto): discounting by p_auto would let
// mass-reporting a target's likely defenders silence their reports
// (report-the-reporters), and the definition would be recursive. Only a
// human-confirmed false-report burn or uphold ever reduces a reporter's
// future report weight.
func ReporterWeight(adjPenalty, inboundTrust float64) float64 {
	return (1 - clamp01(adjPenalty)) * math.Log(1+math.Max(0, inboundTrust))
}

// jaccard is the overlap of two outbound follow sets. Jaccard(∅,∅) = 0 by
// convention here, never 1: two accounts with empty outbound follow sets
// share no evidence of coordination and must NOT be linked just because
// both sets are empty.
func jaccard(a, b map[string]bool) float64 {
	intersection := 0
	for x := range a {
		if b[x] {
			intersection++
		}
	}
	union := len(a) + len(b) - intersection
	if union == 0 {
		return 0
	}
	return float64(intersection) / float64(union)
}

// ClusterReporters returns the connected components of the reporter link
// graph (trust-and-reach §4): reporters R1, R2 are linked iff either
// follows the other, or the Jaccard overlap of their outbound follow sets
// >= jaccardThreshold. Deterministic output: components sorted by their
// lexicographically-smallest member, members sorted within each component.
// `reporters` is deduped defensively.
func ClusterReporters(reporters []string, follows map[string][]string, jaccardThreshold float64) [][]string {
	nodes := make([]string, 0, len(reporters))
	seen := map[string]bool{}
	for _, r := range reporters {
		if seen[r] {
			continue
		}
		seen[r] = true
		nodes = append(nodes, r)
	}
	followSets := make(map[string]map[string]bool, len(nodes))
	for _, r := range nodes {
		followSets[r] = toSet(follows[r])
	}

	// Union-find over the deduped reporter list.
	parent := make(map[string]string, len(nodes))
	for _, r := range nodes {
		parent[r] = r
	}
	var find func(string) string
	find = func(x string) string {
		if parent[x] != x {
			parent[x] = find(parent[x])
		}
		return parent[x]
	}
	union := func(a, b string) {
		ra, rb := find(a), find(b)
		if ra != rb {
			parent[ra] = rb
		}
	}
	for i := 0; i < len(nodes); i++ {
		for j := i + 1; j < len(nodes); j++ {
			a, b := nodes[i], nodes[j]
			linked := followSets[a][b] || followSets[b][a] ||
				jaccard(followSets[a], followSets[b]) >= jaccardThreshold
			if linked {
				union(a, b)
			}
		}
	}

	groups := map[string][]string{}
	roots := []string{}
	for _, r := range nodes {
		root := find(r)
		if _, ok := groups[root]; !ok {
			roots = append(roots, root)
		}
		groups[root] = append(groups[root], r)
	}
	components := make([][]string, 0, len(groups))
	for _, root := range roots {
		members := groups[root]
		sort.Strings(members)
		components = append(components, members)
	}
	sort.Slice(components, func(i, j int) bool {
		return components[i][0] < components[j][0]
	})
	return components
}

// ReportMass is the diversity-weighted report mass (trust-and-reach §4):
// each cluster contributes only its maximum member weight — volume inside a
// cluster adds nothing. A reporter absent from `weights` counts as 0.
func ReportMass(clusters [][]string, weights map[string]float64) float64 {
	mass := 0.0
	for _, cluster := range clusters {
		max := 0.0
		for _, member := range cluster {
			if w := weights[member]; w > max {
				max = w
			}
		}
		mass += max
	}
	return mass
}

// AutoPenaltyWith is p_auto = min(cap, impact × max(0, mass))
// (trust-and-reach §4) with explicit constants — the shape the vector
// exercises and the instance-configurable form.
func AutoPenaltyWith(mass, impact, cap float64) float64 {
	return math.Min(cap, impact*math.Max(0, mass))
}

// AutoPenalty applies AutoPenaltyWith with this instance's published
// constants.
func AutoPenalty(mass float64) float64 {
	return AutoPenaltyWith(mass, ReportImpact, ReportAutoCap)
}

// StandingFrom is standing = (1 - p_auto) × (1 - p_adj) (protocol §9.3),
// each clamped to [0,1].
func StandingFrom(pAuto, pAdj float64) float64 {
	return (1 - clamp01(pAuto)) * (1 - clamp01(pAdj))
}
